transformer_engine/pytorch/utils.py

```diff
@@ -251,3 +252,52 @@ def get_cudnn_version() -> Tuple[int, int, int]:
     major, encoded_version = divmod(encoded_version, major_version_magnitude)
     minor, patch = divmod(encoded_version, 100)
     return (major, minor, patch)
+
+
+def canonicalize_device(device: Optional[torch.device | str]) -> torch.device:
+    """Canonicalize PyTorch device
+
+    If `None`, then returns the default CUDA device.
+
+    """
+    if device is None:
+        # Use default CUDA device
+        device = torch.get_default_device()
+        if device.type != "cuda":
+            device = torch.device("cuda", torch.cuda.current_device())
+    elif not isinstance(device, torch.device):
+        device = torch.device(device)
+    if device.type == "cuda" and device.index is None:
+        device = torch.device("cuda", torch.cuda.current_device())
+    return device
+
+
+def canonicalize_dtype(dtype: Optional[torch.dtype]) -> torch.dtype:
+    """Canonicalize PyTorch datatype
+
+    If `None`, then returns the default PyTorch datatype.
+
+    """
+    if dtype is None:
+        # Use default dtype
+        dtype = torch.get_default_dtype()
+    return dtype
+
+
+def devices_match(device1: torch.device, device2: torch.device) -> bool:
+    """Whether two devices are the same"""
+    device1 = torch.device(device1)
+    device2 = torch.device(device2)
+    if device1.type != device2.type:
+        return False
+    if device1.type == "cuda":
+        index1 = device1.index
+        index2 = device2.index
+        if index1 == index2:
+            return True
+        if index1 is None:
+            index1 = torch.cuda.current_device()
+        if index2 is None:
+            index2 = torch.cuda.current_device()
+        return index1 == index2
```

Comment:
```suggestion
```

Reply:
Anything wrong with this? Pytorch checks `type` and `index` internally.

Reply:
Unfortunately I observe that `torch.device("cuda") == torch.device("cuda:0")` is `False`.

Reply:
Got it, and what about returning `a.index == b.index and a.type == b.type`?

If the device is explicitly not initialized with index (e.g. "cuda"), it's left as `None`?

Reply:
I think it's safer to interpret `device="cuda"` as "the default GPU" rather than "any GPU". For example:
```python
x = torch.zeros(1, device="cuda:1")

torch.cuda.set_device(0)
mydevice = torch.device("cuda")
if not devices_match(x.device, mydevice):
    x = x.to(device=mydevice)
y = torch.zeros(1, device=mydevice)
myfunc(x, y)  # Error: buffers are on different GPUs
```


I suppose this is not critical right now since TE doesn't have good multi-GPU support within a single process (i.e. we only run on the default GPU), but better to get the semantics right.

Reply:
Yep, indexing as current device index is best